File: CSharp/SpaceVIL/SpaceVIL/Items/Label.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Drawing;
using SpaceVIL.Core;
using SpaceVIL.Common;
using SpaceVIL.Decorations;

namespace SpaceVIL
{
    /// <summary>
    /// Label is is the basic implementation of a user interface non-editable text item. 
    /// Label has multiline text support.
    /// <para/> Supports all events except drag and drop.
    /// <para/> By default, Label is stretched to all available space in the container.
    /// </summary>
    public class Label : Prototype, IVLayout
    {
        static int count = 0;
        private List<TextLine> _textObjects;
        private bool _init = false;

        /// <summary>
        /// Property to enable or disable mouse events (hover, click, press, scroll).
        /// <para/> True: Label can receive mouse events. False: cannot receive mouse events.
        /// <para/> Default: True.
        /// </summary>
        public bool IsHover = true;

        protected internal override bool GetHoverVerification(float xpos, float ypos)
        {
            if (IsHover)
            {
                return base.GetHoverVerification(xpos, ypos);
            }
            return false;
        }

        /// <summary>
        /// Default Label constructor. 
        /// </summary>
        public Label()
        {
            SetItemName("Label_" + count);
            count++;
            _textObjects = new List<TextLine>();
            _textObjects.Add(new TextLine());

            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.Label)));
            IsFocusable = false;
        }

        /// <summary>
        /// Constructs a Label with text.
        /// </summary>
        /// <param name="text">Label text.</param>
        public Label(String text) : this()
        {
            SetText(text);
        }
        /// <summary>
        /// Constructs a Label with text and with the ability 
        /// to enable or disable mouse events.
        /// </summary>
        /// <param name="text">Label text.</param>
        /// <param name="hover">True: Label can receive mouse events. 
        /// False: cannot receive mouse events.</param>
        public Label(String text, bool hover) : this(text)
        {
            SetText(text);
            IsHover = hover;
        }

        /// <summary>
        /// Setting alignment of Label text. 
        /// Combines with alignment by vertically (Top, VCenter, Bottom) and horizontally (Left, HCenter, Right). 
        /// </summary>
        /// <param name="alignment">Text alignment as SpaceVIL.Core.ItemAlignment.</param>
        public void SetTextAlignment(ItemAlignment alignment)
        {
            foreach (TextLine tl in _textObjects)
            {
                tl.SetTextAlignment(alignment);
            }
            UpdateLayout();
        }
        /// <summary>
        /// Setting alignment of an Label text. 
        /// Combines with alignment by vertically (Top, VCenter, Bottom) and horizontally (Left, HCenter, Right). 
        /// </summary>
        /// <param name="alignment">Text alignment as sequence of SpaceVIL.Core.ItemAlignment.</param>
        public void SetTextAlignment(params ItemAlignment[] alignment)
        {
            // ItemAlignment common = alignment.ElementAt(0);
            // if (alignment.Length > 1)
            // {
            //     for (int i = 1; i < alignment.Length; i++)
            //     {
            //         common |= alignment.ElementAt(i);
            //     }
            // }
            // SetTextAlignment(common);
            SetTextAlignment(BaseItemStatics.ComposeFlags(alignment));
        }
        /// <summary>
        /// Getting alignment of a Label text. 
        /// </summary>
        /// <returns>Text alignment as SpaceVIL.Core.ItemAlignment.</returns>
        public ItemAlignment GetTextAlignment()
        {
            return _textObjects[0].GetTextAlignment();
        }
        /// <summary>
        /// Setting indents for the text to offset text relative to Label.
        /// </summary>
        /// <param name="margin">Indents as SpaceVIL.Decorations.Indents.</param>
        public void SetTextMargin(Indents margin)
        {
            foreach (TextLine tl in _textObjects)
            {
                tl.SetMargin(margin);
            }
            UpdateLayout();
        }
        /// <summary>
        /// Setting indents for the text to offset text relative to Label.
        /// </summary>
        /// <param name="left">Indent on the left.</param>
        /// <param name="top">Indent on the top.</param>
        /// <param name="right">Indent on the right.</param>
        /// <param name="bottom">Indent on the bottom.</param>
        public void SetTextMargin(int left = 0, int top = 0, int right = 0, int bottom = 0)
        {
            SetTextMargin(new Indents(left, top, right, bottom));
        }
        /// <summary>
        /// Getting indents of the text.
        /// </summary>
        /// <returns>Indents as SpaceVIL.Decorations.Indents.</returns>
        public Indents GetTextMargin()
        {
            return _textObjects[0].GetMargin();
        }

        /// <summary>
        /// Setting font of the text.
        /// </summary>
        /// <param name="font">Font as System.Drawing.Font.</param>
        public void SetFont(Font font)
        {
            foreach (TextLine tl in _textObjects)
            {
                tl.SetFont(font);
            }
            UpdateLayout();
        }
        /// <summary>
        /// Setting font size of the text.
        /// </summary>
        /// <param name="size">New size of the font.</param>
        public void SetFontSize(int size)
        {
            foreach (TextLine tl in _textObjects)
            {
                tl.SetFontSize(size);
            }
            UpdateLayout();
        }
        /// <summary>
        /// Setting font style of the text.
        /// </summary>
        /// <param name="style">New font style as System.Drawing.FontStyle.</param>
        public void SetFontStyle(FontStyle style)
        {
            foreach (TextLine tl in _textObjects)
            {
                tl.SetFontStyle(style);
            }
        }
        /// <summary>
        /// Setting new font family of the text.
        /// </summary>
        /// <param name="fontFamily">New font family as System.Drawing.FontFamily.</param>
        public void SetFontFamily(FontFamily fontFamily)
        {
            foreach (TextLine tl in _textObjects)
            {
                tl.SetFontFamily(fontFamily);
            }
        }
        /// <summary>
        /// Getting the current font of the text.
        /// </summary>
        /// <returns>Font as System.Drawing.Font.</returns>
        public Font GetFont()
        {
            return _textObjects[0].GetFont();
        }

        /// <summary>
        /// Setting the text.
        /// </summary>
        /// <param name="text">Text as System.String.</param>
        public void SetText(String text)
        {
            if (text == null)
            {
                text = "";
            }

            if (_textObjects.Count > 1)
            {
                while (_textObjects.Count > 1)
                {
                    if (_init)
                    {
                        RemoveItem(_textObjects[1]);
                    }
                    _textObjects.RemoveAt(1);
                }
            }

            string[] line = text.Split('\n');
            string s;

            s = line[0].TrimEnd('\r');
            _textObjects[0].SetItemText(s);

            int inc = 0;
            for (int i = 1; i < line.Length; i++)
            {
                inc++;
                s = line[i].TrimEnd('\r');

                TextLine te = new TextLine();
                if (_init)
                {
                    AddItem(te);
                }

                te.SetItemText(s);

                _textObjects.Insert(inc, te);
            }

            SetForeground(GetForeground());
            SetTextAlignment(_textObjects[0].GetTextAlignment());
            SetTextMargin(_textObjects[0].GetMargin());
            SetFont(GetFont());
        }

        /// <summary>
        /// Setting the text.
        /// </summary>
        /// <param name="text">Text as System.Object.</param>
        public virtual void SetText(object text)
        {
            SetText(text.ToString());
        }

        /// <summary>
        /// Getting the current text of the Label.
        /// </summary>
        /// <returns>Text as System.String.</returns>
        public String GetText()
        {
            StringBuilder sb = new StringBuilder();
            if (_textObjects == null)
            {
                return "";
            }
            if (_textObjects.Count == 1)
            {
                sb.Append(_textObjects[0].GetText());
            }
            else
            {
                for (int i = 0; i < _textObjects.Count - 1; i++)
                {
                    sb.Append(_textObjects[i].GetText());
                    sb.Append("\n");
                }
                sb.Append(_textObjects[_textObjects.Count - 1].GetText());
            }
            return sb.ToString();
        }

        /// <summary>
        /// Setting text color of a Label.
        /// </summary>
        /// <param name="color">Text color as System.Drawing.Color.</param>
        public void SetForeground(Color color)
        {
            foreach (TextLine tl in _textObjects)
            {
                tl.SetForeground(color);
            }
        }
        /// <summary>
        /// Setting text color of a Label in byte RGB format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0 - 255)</param>
        /// <param name="g">Green color component. Range: (0 - 255)</param>
        /// <param name="b">Blue color component. Range: (0 - 255)</param>
        public void SetForeground(int r, int g, int b)
        {
            foreach (TextLine tl in _textObjects)
            {
                tl.SetForeground(r, g, b);
            }
        }
        /// <summary>
        /// Setting text color of a Label in byte RGBA format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0 - 255)</param>
        /// <param name="g">Green color component. Range: (0 - 255)</param>
        /// <param name="b">Blue color component. Range: (0 - 255)</param>
        /// <param name="a">Alpha color component. Range: (0 - 255)</param>
        public void SetForeground(int r, int g, int b, int a)
        {
            foreach (TextLine tl in _textObjects)
            {
                tl.SetForeground(r, g, b, a);
            }
        }
        /// <summary>
        /// Setting text color of a Label in float RGB format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0.0f - 1.0f)</param>
        /// <param name="g">Green color component. Range: (0.0f - 1.0f)</param>
        /// <param name="b">Blue color component. Range: (0.0f - 1.0f)</param>
        public void SetForeground(float r, float g, float b)
        {
            foreach (TextLine tl in _textObjects)
            {
                tl.SetForeground(r, g, b);
            }
        }
        /// <summary>
        /// Setting text color of a Label in float RGBA format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0.0f - 1.0f)</param>
        /// <param name="g">Green color component. Range: (0.0f - 1.0f)</param>
        /// <param name="b">Blue color component. Range: (0.0f - 1.0f)</param>
        /// <param name="a">Alpha color component. Range: (0.0f - 1.0f)</param>
        public void SetForeground(float r, float g, float b, float a)
        {
            foreach (TextLine tl in _textObjects)
            {
                tl.SetForeground(r, g, b, a);
            }
        }
        /// <summary>
        /// Getting current text color.
        /// </summary>
        /// <returns>Text color as System.Drawing.Color.</returns>
        public Color GetForeground()
        {
            return _textObjects[0].GetForeground();
        }
        /// <summary>
        /// Getting the text width (useful when you need resize Label by text width).
        /// </summary>
        /// <returns>Text width.</returns>
        public int GetTextWidth()
        {
            int wdt = _textObjects[0].GetWidth();
            for (int i = 1; i < _textObjects.Count; i++)
            {
                int w = _textObjects[i].GetWidth();
                if (w > wdt)
                {
                    wdt = w;
                }
            }
            return wdt;
        }
        /// <summary>
        /// Getting the text height (useful when you need resize Label by text height).
        /// </summary>
        /// <returns>Text height.</returns>
        public int GetTextHeight()
        {
            return GetLineY(_textObjects.Count);
        }

        private int GetLineY(int num)
        {
            int minLineSpacer = _textObjects[0].GetFontDims().lineSpacer; //[0];
            int lineHeight = _textObjects[0].GetHeight();
            return (lineHeight + minLineSpacer) * num;
        }
        /// <summary>
        /// Setting Label width. If the value is greater/less than the maximum/minimum 
        /// value of the width, then the width becomes equal to the maximum/minimum value.
        /// </summary>
        /// <param name="width"> Width of the Label. </param>
        public override void SetWidth(int width)
        {
            base.SetWidth(width);
            TextLine txtObj = _textObjects[0];
            int _cursorXMax = GetWidth() - GetPadding().Left - GetPadding().Right - txtObj.GetMargin().Left
                    - txtObj.GetMargin().Right;

            foreach (TextLine tl in _textObjects)
            {
                tl.SetAllowWidth(_cursorXMax);
                tl.CheckXShift(_cursorXMax); // ???
            }
        }

        /// <summary>
        /// Initializing all elements in the Label.
        /// <para/> Notice: This method is mainly for overriding only. SpaceVIL calls 
        /// this method if necessary and no need to call it manually.
        /// </summary>
        public override void InitElements()
        {
            foreach (TextLine tl in _textObjects)
            {
                AddItem(tl);
            }
            _init = true;
        }

        /// <summary>
        /// Setting style of the Label.
        /// </summary>
        /// <param name="style">Style as SpaceVIL.Decorations.Style.</param>
        public override void SetStyle(Style style)
        {
            if (style == null)
            {
                return;
            }
            base.SetStyle(style);

            SetFont(style.Font);
            SetForeground(style.Foreground);
            SetTextAlignment(style.TextAlignment);
        }
        /// <summary>
        /// Updating all children positions (implementation of SpaceVIL.Core.IVLayout).
        /// </summary>
        public void UpdateLayout()
        {
            ItemAlignment alignment = GetTextAlignment();
            int globalYShift = 0;
            if (alignment.HasFlag(ItemAlignment.Bottom))
            {
                globalYShift = -(GetTextHeight() - GetLineY(1));
            }
            else if (alignment.HasFlag(ItemAlignment.VCenter))
            {
                globalYShift = -((GetTextHeight() - GetLineY(1)) / 2);
            }
            int inc = 0;
            foreach (TextLine tl in _textObjects)
            {
                tl.SetLineYShift(GetLineY(inc) + globalYShift);
                inc++;
            }
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/Label.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Common.DefaultsService;
import com.spvessel.spacevil.Core.InterfaceVLayout;
import com.spvessel.spacevil.Decorations.Indents;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.ItemAlignment;

import java.awt.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Label is is the basic implementation of a user interface non-editable text
 * item. Label has multiline text support.
 * <p>
 * Supports all events except drag and drop.
 * <p>
 * By default, Label is stretched to all available space in the container.
 */
public class Label extends Prototype implements InterfaceVLayout {
    private static int count = 0;
    private List<TextLine> _textObjects;
    private boolean _init = false;
    /**
     * Property to enable or disable mouse events (hover, click, press, scroll).
     * <p>
     * True: Label can receive mouse events. False: cannot receive mouse events.
     * <p>
     * Default: True.
     */
    public boolean isHover = true;

    @Override
    protected boolean getHoverVerification(float xpos, float ypos) {
        if (isHover) {
            return super.getHoverVerification(xpos, ypos);
        }
        return false;
    }

    /**
     * Default Label constructor.
     */
    public Label() {
        setItemName("Label_" + count);
        count++;
        _textObjects = new ArrayList<>();
        _textObjects.add(new TextLine());
        setStyle(DefaultsService.getDefaultStyle(Label.class));
        isFocusable = false;
    }

    /**
     * Constructs a Label with text.
     * 
     * @param text Label text.
     */
    public Label(String text) {
        this();
        setText(text);
    }

    /**
     * Constructs a Label with text and with the ability to enable or disable mouse
     * events.
     * 
     * @param text  Label text.
     * @param hover True: Label can receive mouse events. False: cannot receive
     *              mouse events.
     */
    public Label(String text, boolean hover) {
        this(text);
        isHover = hover;
    }

    /**
     * Setting alignment of a Label text. Combines with alignment by vertically
     * (TOP, VCENTER, BOTTOM) and horizontally (LEFT, HCENTER, RIGHT).
     * 
     * @param alignment Text alignment as com.spvessel.spacevil.Flags.ItemAlignment.
     */
    public void setTextAlignment(ItemAlignment... alignment) {
        setTextAlignment(Arrays.asList(alignment));
    }

    /**
     * Setting alignment of a Label text. Combines with alignment by vertically
     * (TOP, VCENTER, BOTTOM) and horizontally (LEFT, HCENTER, RIGHT).
     * 
     * @param alignment Text alignment as List of
     *                  com.spvessel.spacevil.Flags.ItemAlignment.
     */
    public void setTextAlignment(List<ItemAlignment> alignment) {
        for (TextLine tl : _textObjects) {
            tl.setTextAlignment(alignment);
        }
        updateLayout();
    }

    /**
     * Getting alignment of a Label text.
     * 
     * @return Text alignment as List of com.spvessel.spacevil.Flags.ItemAlignment.
     */
    public List<ItemAlignment> getTextAlignment() {
        return _textObjects.get(0).getTextAlignment();
    }

    /**
     * Setting indents for the text to offset text relative to Label.
     * 
     * @param margin Indents as com.spvessel.spacevil.Decorations.Indents.
     */
    public void setTextMargin(Indents margin) {
        for (TextLine tl : _textObjects) {
            tl.setMargin(margin);
        }
        updateLayout();
    }

    /**
     * Setting indents for the text to offset text relative to Label.
     * 
     * @param left   Indent on the left.
     * @param top    Indent on the top.
     * @param right  Indent on the right.
     * @param bottom Indent on the bottom.
     */
    public void setTextMargin(int left, int top, int right, int bottom) {
        setTextMargin(new Indents(left, top, right, bottom));
    }

    /**
     * Getting indents of the text.
     * 
     * @return Indents as com.spvessel.spacevil.Decorations.Indents.
     */
    public Indents getTextMargin() {
        return _textObjects.get(0).getMargin();
    }

    /**
     * Setting font of the text.
     * 
     * @param font Font as java.awt.Font.
     */
    public void setFont(Font font) {
        for (TextLine tl : _textObjects) {
            tl.setFont(font);
        }
        updateLayout();
    }

    /**
     * Setting font size of the text.
     * 
     * @param size New size of the font.
     */
    public void setFontSize(int size) {
        for (TextLine tl : _textObjects) {
            tl.setFontSize(size);
        }
        updateLayout();
    }

    /**
     * Setting font style of the text.
     * 
     * @param style New font style (from java.awt.Font package).
     */
    public void setFontStyle(int style) {
        for (TextLine tl : _textObjects) {
            tl.setFontStyle(style);
        }
    }

    /**
     * Setting new font family of the text.
     * 
     * @param fontFamily New font family name.
     */
    public void setFontFamily(String font_family) {
        for (TextLine tl : _textObjects) {
            tl.setFontFamily(font_family);
        }
    }

    /**
     * Getting the current font of the text.
     * 
     * @return Font as java.awt.Font.
     */
    public Font getFont() {
        return _textObjects.get(0).getFont();
    }

    /**
     * Setting the text.
     * 
     * @param text Text as java.lang.String.
     */
    public void setText(String text) {
        if (text == null) {
            text = "";
        }

        if (_textObjects.size() > 1) {
            while (_textObjects.size() > 1) {
                if (_init) {
                    removeItem(_textObjects.get(1));
                }
                _textObjects.remove(1);
            }
        }

        String[] line = text.split("\n", -1);
        String s;

        s = line[0].replaceAll("\r", "");
        _textObjects.get(0).setItemText(s);

        int inc = 0;
        for (int i = 1; i < line.length; i++) {
            inc++;
            s = line[i].replaceAll("\r", "");

            TextLine te = new TextLine();
            if (_init) {
                addItem(te);
            }

            te.setItemText(s);

            _textObjects.add(inc, te);
        }

        setForeground(getForeground());
        setTextAlignment(_textObjects.get(0).getTextAlignment());
        setTextMargin(_textObjects.get(0).getMargin());
        setFont(getFont());
    }

    /**
     * Getting the current text of the Label.
     * 
     * @return Text as java.lang.String.
     */
    public String getText() {
        StringBuilder sb = new StringBuilder();
        if (_textObjects == null) {
            return "";
        }
        if (_textObjects.size() == 1) {
            sb.append(_textObjects.get(0).getText());
        } else {
            for (int i = 0; i < _textObjects.size() - 1; i++) {
                sb.append(_textObjects.get(i).getText());
                sb.append("\n");
            }
            sb.append(_textObjects.get(_textObjects.size() - 1).getText());
        }
        return sb.toString();
    }

    /**
     * Setting text color of a Label.
     * 
     * @param color Text color as java.awt.Color.
     */
    public void setForeground(Color color) {
        for (TextLine tl : _textObjects) {
            tl.setForeground(color);
        }
    }

    /**
     * Setting text color of a Label in byte RGB format.
     * 
     * @param r Red color component. Range: (0 - 255)
     * @param g Green color component. Range: (0 - 255)
     * @param b Blue color component. Range: (0 - 255)
     */
    public void setForeground(int r, int g, int b) {
        for (TextLine tl : _textObjects) {
            tl.setForeground(r, g, b);
        }
    }

    /**
     * Setting text color of an item in byte RGBA format.
     * 
     * @param r Red color component. Range: (0 - 255)
     * @param g Green color component. Range: (0 - 255)
     * @param b Blue color component. Range: (0 - 255)
     * @param a Alpha color component. Range: (0 - 255)
     */
    public void setForeground(int r, int g, int b, int a) {
        for (TextLine tl : _textObjects) {
            tl.setForeground(r, g, b, a);
        }
    }

    /**
     * Setting text color of a Label in float RGB format.
     * 
     * @param r Red color component. Range: (0.0f - 1.0f)
     * @param g Green color component. Range: (0.0f - 1.0f)
     * @param b Blue color component. Range: (0.0f - 1.0f)
     */
    public void setForeground(float r, float g, float b) {
        for (TextLine tl : _textObjects) {
            tl.setForeground(r, g, b);
        }
    }

    /**
     * Setting text color of a Label in float RGBA format.
     * 
     * @param r Red color component. Range: (0.0f - 1.0f)
     * @param g Green color component. Range: (0.0f - 1.0f)
     * @param b Blue color component. Range: (0.0f - 1.0f)
     * @param a Alpha color component. Range: (0.0f - 1.0f)
     */
    public void setForeground(float r, float g, float b, float a) {
        for (TextLine tl : _textObjects) {
            tl.setForeground(r, g, b, a);
        }
    }

    /**
     * Getting current text color.
     * 
     * @return Text color as as java.awt.Color.
     */
    public Color getForeground() {
        return _textObjects.get(0).getForeground();
    }

    /**
     * Getting the text width (useful when you need resize Label by text width).
     * 
     * @return Text width.
     */
    public int getTextWidth() {
        int wdt = _textObjects.get(0).getWidth();
        for (int i = 1; i < _textObjects.size(); i++) {
            int w = _textObjects.get(i).getWidth();
            if (w > wdt) {
                wdt = w;
            }
        }
        return wdt;
    }

    /**
     * Getting the text height (useful when you need resize Label by text height).
     * 
     * @return Text height.
     */
    public int getTextHeight() {
        return getLineY(_textObjects.size());
    }

    private int getLineY(int num) {
        int minLineSpacer = _textObjects.get(0).getFontDims().lineSpacer; // [0];
        int lineHeight = _textObjects.get(0).getHeight();
        return (lineHeight + minLineSpacer) * num;
    }

    /**
     * Setting Label width. If the value is greater/less than the maximum/minimum
     * value of the width, then the width becomes equal to the maximum/minimum
     * value.
     * 
     * @param width Width of the Label.
     */
    @Override
    public void setWidth(int width) {
        super.setWidth(width);
        TextLine txtObj = _textObjects.get(0);
        int _cursorXMax = getWidth() - getPadding().left - getPadding().right - txtObj.getMargin().left
                - txtObj.getMargin().right;

        for (TextLine tl : _textObjects) {
            tl.setAllowWidth(_cursorXMax);
            tl.checkXShift(_cursorXMax); // ???
        }
    }

    /**
     * Initializing all elements in the Label.
     * <p>
     * Notice: This method is mainly for overriding only. SpaceVIL calls this method
     * if necessary and no need to call it manually.
     */
    @Override
    public void initElements() {
        for (TextLine tl : _textObjects) {
            addItem(tl);
        }
        _init = true;
    }

    /**
     * Setting style of the Label.
     * 
     * @param style Style as com.spvessel.spacevil.Decorations.Style.
     */
    @Override
    public void setStyle(Style style) {
        if (style == null) {
            return;
        }
        super.setStyle(style);
        setFont(style.font);
        setForeground(style.foreground);
        setTextAlignment(style.textAlignment);
    }

    /**
     * Updating all children positions (implementation of
     * com.spvessel.spacevil.Core.InterfaceVLayout).
     */
    @Override
    public void updateLayout() {
        // updateLinesYShifts
        List<ItemAlignment> alignment = getTextAlignment();
        int globalYShift = 0;
        if (alignment.contains(ItemAlignment.BOTTOM)) {
            globalYShift = -(getTextHeight() - getLineY(1));
        } else if (alignment.contains(ItemAlignment.VCENTER)) {
            globalYShift = -((getTextHeight() - getLineY(1)) / 2);
        }
        int inc = 0;
        for (TextLine tl : _textObjects) {
            tl.setLineYShift(getLineY(inc) + globalYShift);
            inc++;
        }
    }
}